handler.go

```diff
@@ -0,0 +1,318 @@
+}
+
+// makeServerStream starts server-side streaming rpc
+func makeServerStream(ctx context.Context, stream *yarpctransport.ClientStream,
+	reqSupplier streamRequestSupplier, resHandler streamResponseHandler) error {
+
+	req, err := reqSupplier()
+	if err != nil && err != io.EOF {
+		return err
+	}
+	if err == nil {
+		// verify there is no second request
+		_, err = reqSupplier()
+		if err == nil {
+			return fmt.Errorf("Request data contains more than 1 message for server-streaming RPC")
+		} else if err != io.EOF {
+			return err
+		}
+	}
+
+	err = sendStreamMessage(ctx, stream, req)
+	if err != nil {
+		return err
+	}
+
+	for err == nil {
+		var resBody []byte
+		resBody, err = receiveStreamMessage(ctx, stream)
+		if err != nil {
+			break
+		}
+		err = resHandler(resBody)
+	}
+
+	if err == io.EOF {
+		return nil
+	}
+	return err
+}
+
+// makeClientStream starts client-side streaming rpc
+func makeClientStream(ctx context.Context, stream *yarpctransport.ClientStream,
+	reqSupplier streamRequestSupplier, resHandler streamResponseHandler) error {
+
+	var err error
+	for err == nil {
+		var reqBody []byte
+		reqBody, err = reqSupplier()
+		if err != nil {
+			break
+		}
+		err = sendStreamMessage(ctx, stream, reqBody)
+	}
+
+	if err == io.EOF {
+		err = closeSendStream(ctx, stream)
+	}
+	if err != nil {
+		return err
+	}
+
+	res, err := receiveStreamMessage(ctx, stream)
+	if err != nil {
+		return err
+	}
+	return resHandler(res)
+}
+
+// makeBidiStream starts bi-directional streaming rpc
+func makeBidiStream(ctx context.Context, stream *yarpctransport.ClientStream,
+	reqSupplier streamRequestSupplier, resHandler streamResponseHandler) error {
+
+	var wg sync.WaitGroup
+	var sendErr error
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	wg.Add(1)
+	// start go routine to concurrently send stream messages
+	go func() {
+		defer wg.Done()
+
+		var err error
+		for err == nil {
+			var reqBody []byte
+			reqBody, err = reqSupplier()
+			if err == io.EOF {
+				err = closeSendStream(ctx, stream)
+				break
+			}
+			if err != nil {
+				cancel()
+				break
+			}
+
+			err = sendStreamMessage(ctx, stream, reqBody)
+		}
+
+		if err != nil {
+			sendErr = err
+		}
+	}()
+
+	var err error
+	for err == nil {
+		var resBody []byte
+		resBody, err = receiveStreamMessage(ctx, stream)
+		if err != nil {
+			break
+		}
+		err = resHandler(resBody)
+	}
+
+	cancel()
+	wg.Wait()
+
+	if sendErr != nil && sendErr != io.EOF {
+		err = sendErr
+	}
+
+	if err == io.EOF {
+		return nil
+	}
+	return err
+}
+
+// sendStreamMessage sends the stream message using message body provided
+func sendStreamMessage(ctx context.Context, stream *yarpctransport.ClientStream, msgBody []byte) error {
+	req := &yarpctransport.StreamMessage{Body: ioutil.NopCloser(bytes.NewReader(msgBody))}
+	if err := stream.SendMessage(ctx, req); err != nil {
+		return fmt.Errorf("Failed while sending stream request: %v", err)
+	}
+	return nil
+}
```

Comment:
Should we also print the message we are sending? This can be very helpful in a streaming mode IMO, without it it will be hard to match responses with requests.

Reply:
We shouldn't do it, it would look confusing when streaming is used with interactive mode (STDIN) as user will see the request twice

Reply:
Adding a TODO for this feature